File: server/services/llm.py
from anthropic import Anthropic
from openai import OpenAI

from server.config import (
    AI_MODEL,
    AI_PROVIDER,
    ANTHROPIC_API_KEY,
    GROQ_API_KEY,
    OLLAMA_BASE_URL,
    OPENAI_API_KEY,
)


def _anthropic_chat(system_prompt, messages, max_tokens):
    if not ANTHROPIC_API_KEY:
        raise RuntimeError("ANTHROPIC_API_KEY is not configured")

    client = Anthropic(api_key=ANTHROPIC_API_KEY)
    response = client.messages.create(
        model=AI_MODEL,
        max_tokens=max_tokens,
        system=system_prompt,
        messages=messages,
    )
    return "".join(block.text for block in response.content if block.type == "text")


def _openai_compatible_client():
    if AI_PROVIDER == "groq":
        if not GROQ_API_KEY:
            raise RuntimeError("GROQ_API_KEY is not configured")
        return OpenAI(api_key=GROQ_API_KEY, base_url="https://api.groq.com/openai/v1")

    if AI_PROVIDER == "ollama":
        return OpenAI(api_key="ollama", base_url=OLLAMA_BASE_URL)

    if AI_PROVIDER == "openai":
        if not OPENAI_API_KEY:
            raise RuntimeError("OPENAI_API_KEY is not configured")
        return OpenAI(api_key=OPENAI_API_KEY)

    raise RuntimeError(f"Unknown AI_PROVIDER '{AI_PROVIDER}'")


def _openai_compatible_chat(system_prompt, messages, max_tokens):
    client = _openai_compatible_client()
    try:
        response = client.chat.completions.create(
            model=AI_MODEL,
            max_tokens=max_tokens,
            messages=[{"role": "system", "content": system_prompt}, *messages],
        )
    except Exception as exc:
        raise RuntimeError(f"{AI_PROVIDER} request failed: {exc}") from exc
    return response.choices[0].message.content or ""


def chat(system_prompt, messages, max_tokens=1024):
    if AI_PROVIDER == "anthropic":
        return _anthropic_chat(system_prompt, messages, max_tokens)
    return _openai_compatible_chat(system_prompt, messages, max_tokens)

File: functions/_lib/llm.js
async function anthropicChat(env, systemPrompt, messages, maxTokens) {
  if (!env.ANTHROPIC_API_KEY) throw new Error("ANTHROPIC_API_KEY is not configured");

  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": env.ANTHROPIC_API_KEY,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: env.AI_MODEL || "claude-sonnet-4-6",
      max_tokens: maxTokens,
      system: systemPrompt,
      messages,
    }),
  });
  if (!res.ok) throw new Error(`anthropic request failed: ${await res.text()}`);
  const data = await res.json();
  return data.content.filter((b) => b.type === "text").map((b) => b.text).join("");
}

function providerEndpoint(env, provider) {
  if (provider === "groq") {
    return {
      url: "https://api.groq.com/openai/v1/chat/completions",
      key: env.GROQ_API_KEY,
      model: env.AI_MODEL || "llama-3.3-70b-versatile",
    };
  }
  if (provider === "ollama") {
    return {
      url: `${(env.OLLAMA_BASE_URL || "http://localhost:11434/v1").replace(/\/$/, "")}/chat/completions`,
      key: "ollama",
      model: env.AI_MODEL || "llama3.1",
    };
  }
  if (provider === "openai") {
    return {
      url: "https://api.openai.com/v1/chat/completions",
      key: env.OPENAI_API_KEY,
      model: env.AI_MODEL || "gpt-4o-mini",
    };
  }
  throw new Error(`Unknown AI_PROVIDER '${provider}'`);
}

async function openaiCompatibleChat(env, provider, systemPrompt, messages, maxTokens) {
  const { url, key, model } = providerEndpoint(env, provider);
  if (!key) throw new Error(`${provider} API key is not configured`);

  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model,
      max_tokens: maxTokens,
      messages: [{ role: "system", content: systemPrompt }, ...messages],
    }),
  });
  if (!res.ok) throw new Error(`${provider} request failed: ${await res.text()}`);
  const data = await res.json();
  return data.choices[0].message.content || "";
}

export async function chat(env, systemPrompt, messages, maxTokens = 1024) {
  const provider = (env.AI_PROVIDER || "groq").trim().toLowerCase();
  if (provider === "anthropic") return anthropicChat(env, systemPrompt, messages, maxTokens);
  return openaiCompatibleChat(env, provider, systemPrompt, messages, maxTokens);
}
